core/src/main/java/org/apache/iceberg/deletes/Deletes.java

```diff
@@ -293,15 +310,53 @@ private static class PositionStreamDeleteMarker<T> extends PositionStreamDeleteI
     }
 
     @Override
-    protected CloseableIterator<T> applyDelete(CloseableIterator<T> items) {
-      return CloseableIterator.transform(
-          items,
-          row -> {
-            if (isDeleted(row)) {
-              markDeleted.accept(row);
+    protected CloseableIterator<T> applyDelete(
+        CloseableIterator<T> items, CloseableIterator<Long> deletePositions) {
+
+      return new CloseableIterator<T>() {
+
+        private boolean closed = false;
+
+        @Override
+        public void close() {
+          if (!closed) {
+            try {
+              deletePositions.close();
+            } catch (IOException ignore) {
+              LOG.warn("Error closing delete file", ignore);
             }
-            return row;
-          });
+            try {
+              items.close();
+            } catch (IOException ignore) {
+              LOG.warn("Error closing data file", ignore);
+            }
+            closed = true;
+          }
+        }
+
+        @Override
+        public boolean hasNext() {
```

Comment:
Note: the point of this additional code is to have parity with the other applyDelete method.  

The other one returns a FilterIterator, which auto-closes, as mentioned in description.